### PR TITLE
feat(datasource/npm): cache etag for reuse

### PR DESCRIPTION
**Pull request by rarkins**
## Changes

Adds "soft" expiry package cache concept to npm datasource, reuses etag value and cached result for subsequence requests.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/datasource/npm/get.ts [hunk 1]
```diff
@@ -56,19 +62,54 @@ export async function getDependency(
...
     cacheNamespace,
     packageUrl
   );
-  // istanbul ignore if
   if (cachedResult) {
-    return cachedResult;
+    if (cachedResult.cacheData) {
+      const softExpireAt = new Date(cachedResult.cacheData.softExpireAt);
+      if (softExpireAt > new Date()) {
+        logger.debug('Cached result is not expired - reusing');
```
Comment (rarkins) [inline lib/modules/datasource/npm/get.ts:73]:
```suggestion
        logger.trace('Cached result is not expired - reusing');
```

File: lib/modules/datasource/npm/get.ts [hunk 2]
```diff
@@ -56,19 +62,54 @@ export async function getDependency(
...
+        delete cachedResult.cacheData;
+        return cachedResult;
+      }
+      logger.debug('Cached result is soft expired');
```
Comment (rarkins) [inline lib/modules/datasource/npm/get.ts:77]:
```suggestion
      logger.trace('Cached result is soft expired');
```

File: lib/modules/datasource/npm/get.ts [hunk 3]
```diff
@@ -56,19 +62,54 @@ export async function getDependency(
...
+    } else {
+      logger.debug('Reusing legacy cached result');
```
Comment (rarkins) [inline lib/modules/datasource/npm/get.ts:79]:
```suggestion
      logger.trace('Reusing legacy cached result');
```

File: lib/modules/datasource/npm/get.ts [hunk 4]
```diff
@@ -56,19 +62,54 @@ export async function getDependency(
...
   try {
-    const raw = await http.getJson<NpmResponse>(packageUrl);
+    const options: HttpOptions = {};
+    if (cachedResult?.cacheData?.etag) {
+      logger.debug('Using cached etag');
+      options.headers = { 'If-None-Match': cachedResult.cacheData.etag };
+    }
+    const raw = await http.getJson<NpmResponse>(packageUrl, options);
+    if (cachedResult?.cacheData && raw.statusCode === 304) {
+      logger.debug('Cached data is unchanged and can be reused');
```
Comment (rarkins) [inline lib/modules/datasource/npm/get.ts:101]:
```suggestion
      logger.trace('Cached data is unchanged and can be reused');
```

File: lib/modules/datasource/npm/get.ts [hunk 5]
```diff
@@ -140,7 +178,13 @@ export async function getDependency(
       (whitelistedPublicScopes.includes(packageName.split('/')[0]) ||
         !packageName.startsWith('@'))
     ) {
-      await packageCache.set(cacheNamespace, packageUrl, dep, cacheMinutes);
+      const cacheData = { softExpireAt, etag };
+      await packageCache.set(
+        cacheNamespace,
+        packageUrl,
+        { ...dep, cacheData },
+        hardExpireMinutes
```
Comment (viceice) [inline lib/modules/datasource/npm/get.ts:186]: we should use softExpire if no etag
```suggestion
        etag ? hardExpireMinutes : cacheMinutes
```

File: lib/modules/datasource/npm/get.ts [hunk 6]
```diff
@@ -56,19 +62,54 @@ export async function getDependency(
...
+    } else {
+      logger.trace('Reusing legacy cached result');
+      return cachedResult;
+    }
   }
+  const cacheMinutes = process.env.RENOVATE_CACHE_NPM_MINUTES
+    ? parseInt(process.env.RENOVATE_CACHE_NPM_MINUTES, 10)
+    : 1;
+  const newDate = new Date();
+  newDate.setMinutes(newDate.getMinutes() + cacheMinutes);
```
Comment (viceice) [inline lib/modules/datasource/npm/get.ts:87]: Use luxon

File: lib/modules/datasource/npm/get.ts [hunk 7]
```diff
@@ -56,19 +63,54 @@ export async function getDependency(
...
     packageUrl
   );
-  // istanbul ignore if
   if (cachedResult) {
-    return cachedResult;
+    if (cachedResult.cacheData) {
+      const softExpireAt = DateTime.fromISO(
+        cachedResult.cacheData.softExpireAt
+      );
+      if (softExpireAt > DateTime.local()) {
```
Comment (viceice) [inline lib/modules/datasource/npm/get.ts:75]:
```suggestion
      if (softExpireAt.isValid && softExpireAt > DateTime.local()) {
```
`DateTime.fromISO(undefined) > DateTime.local() === false`
sanity check

File: lib/modules/datasource/npm/get.ts [hunk 8]
```diff
@@ -56,19 +63,54 @@ export async function getDependency(
...
+    } else {
+      logger.trace('Reusing legacy cached result');
+      return cachedResult;
+    }
   }
+  const cacheMinutes = process.env.RENOVATE_CACHE_NPM_MINUTES
+    ? parseInt(process.env.RENOVATE_CACHE_NPM_MINUTES, 10)
+    : 15;
+  const softExpireAt = DateTime.local().plus({ minutes: cacheMinutes }).toISO();
+  const hardExpireMinutes = 24 * 60; // 1 day
```
Comment (viceice) [inline lib/modules/datasource/npm/get.ts:90]: maybe make this an experimental env for fine-tuning?

Comment (rarkins) [reply lib/modules/datasource/npm/get.ts]: I made it an experimental self-hosted config option

**Review by viceice — APPROVED** [review]

**Pull request merged**